File: C/BSTreeFindKthSmallest.c
//Given a binary search tree and a number k, print the kth smallest number in tree.


#include<stdio.h>
#include<stdlib.h>

struct Node
{
  int data;
  struct Node *left, *right;
};

struct Node* newNode(int data)
{
  struct Node* node = (struct Node*)malloc(sizeof(struct Node));
  node->data = data;
  node->left = node->right = NULL;
  return (node);
}

struct Node* insertLevelOrder(int arr[], struct Node* root, int i, int n)
{
  if (i < n)
  {
    struct Node* temp = newNode(arr[i]);
    root = temp;

    root->left = insertLevelOrder(arr, root->left, 2 * i + 1, n);

    root->right = insertLevelOrder(arr, root->right, 2 * i + 2, n);
  }
  return root;
}

struct Node* buildTree(int t[], int n)
{
  struct Node* root = (struct Node*)malloc(sizeof(struct Node));
  root = insertLevelOrder(t, root, 0, n);
  return(root);
}


static int a[10], i=0;
void inorder(struct Node* t1)
{
  if(t1!=NULL)
  {
    inorder(t1->left);
    a[i++]=t1->data;
    inorder(t1->right);
  }
}

int kSmallest(struct Node* t1, int k)
{
  i=0;
  inorder(t1);
  return a[k-1];

}


int main()
{
  int a[10], n, i, k;
  scanf("%d", &n);
  for(i=0;i<n;i++)
    scanf("%d", &a[i]);
  scanf("%d", &k);
  struct Node* root = NULL;
  root = buildTree(a, n);
  printf("%d", kSmallest(root, k));
  return 0;
}

File: Java/BSTreeFindKthSmallest.java
//Given a binary search tree and a number k, print the kth smallest number in tree.

import java.util.Scanner;
class Node
{
  int data; // data used as key value
  Node leftChild;
  Node rightChild;
  public Node()
  {
    data=0;
  }
  public Node(int d)
  {
    data=d;
  }
}

class Main
{
  static void inOrder(Node root)
  {
    if (root != null)
    {
      inOrder(root.leftChild);
      System.out.print(root.data + " ");
      inOrder(root.rightChild);
    }
  }
  static Node insertLevelOrder(int arr[], Node root, int i, int n)
  {
    if (i < n)
    {
      Node temp = new Node(arr[i]);
      root = temp;
      root.leftChild = insertLevelOrder(arr, root.leftChild, 2 * i + 1, n);
      root.rightChild = insertLevelOrder(arr, root.rightChild, 2 * i + 2, n);
    }
    return root;
  }

  static Node buildTree(int t[], int n)
  {
    Node node = null;
    node = insertLevelOrder(t, node, 0, n);
    return node;
  }

  
 static int a[] = new int[10], i=0;
 static void inorder(Node p)
 {
   if(p!=null)
   {
     inorder(p.leftChild);
     a[i++]=p.data;
     inorder(p.rightChild);
   }
 }
 
   static int kSmallest(Node t1, int k)
 {
     inorder(t1);
     if(k==0)
     return a[k];
     else
       return a[k-1];
 
 }

  public static void main(String[] args)
  {
    int a[]=new int[10];
    int n, m, i;
    Scanner s=new Scanner(System.in);
    n = Integer.parseInt(s.nextLine());
    for(i=0;i<n;i++)
      a[i] = Integer.parseInt(s.nextLine());
    i = Integer.parseInt(s.nextLine());
    Node root = null;
    if(n!=0)
      root = buildTree(a, n);
    System.out.println(kSmallest(root, i));
  }
}